pkg/connectorrunner/runner.go

```diff
@@ -1,86 +1,154 @@
 	"go.uber.org/zap"
 	"google.golang.org/protobuf/types/known/durationpb"
 
 	"github.com/conductorone/baton-sdk/internal/connector"
-	"github.com/conductorone/baton-sdk/pkg/connectorstore"
 )
 
 type connectorRunner struct {
-	syncer  sync.Syncer
-	store   connectorstore.Writer
-	manager manager.Manager
+	cw           types.ClientWrapper
+	dbPath       string
+	onDemandMode bool
+	tasks        tasks.Manager
 }
 
-func (c *connectorRunner) shutdown(ctx context.Context) error {
-	logger := ctxzap.Extract(ctx)
+var ErrSigTerm = errors.New("context cancelled by process shutdown")
 
-	err := c.Close()
-	if err != nil {
-		// Explicitly ignoring the error here as it is possible that things have already been closed.
-		logger.Error("error closing connector runner", zap.Error(err))
-	}
+// Run starts a connector and creates a new C1Z file.
+func (c *connectorRunner) Run(ctx context.Context) error {
+	ctx, cancel := context.WithCancelCause(ctx)
+	defer cancel(ErrSigTerm)
+
+	sigChan := make(chan os.Signal, 1)
+	signal.Notify(sigChan, os.Interrupt)
+	go func() {
+		for range sigChan {
+			cancel(ErrSigTerm)
+		}
+	}()
 
-	err = c.manager.SaveC1Z(ctx)
+	err := c.run(ctx)
 	if err != nil {
-		logger.Error("error saving c1z", zap.Error(err))
 		return err
 	}
 
-	err = c.manager.Close(ctx)
+	err = c.Close(ctx)
 	if err != nil {
-		logger.Error("error closing c1z manager", zap.Error(err))
 		return err
 	}
 
 	return nil
 }
 
-// Run starts a connector and creates a new C1Z file.
-func (c *connectorRunner) Run(ctx context.Context) error {
-	ctx, cancel := context.WithCancel(ctx)
-	defer cancel()
-	defer func(c *connectorRunner, ctx context.Context) {
-		err := c.shutdown(ctx)
+func (c *connectorRunner) handleContextCancel(ctx context.Context) error {
+	l := ctxzap.Extract(ctx)
+
+	err := context.Cause(ctx)
+	if err == nil {
+		return nil
+	}
+
+	// The context was cancelled due to an expected end of the process. Swallow the error.
+	if errors.Is(err, ErrSigTerm) {
+		return nil
+	}
+
+	l.Debug("unexpected context cancellation", zap.Error(err))
+	return err
+}
+
+func (c *connectorRunner) run(ctx context.Context) error {
+	l := ctxzap.Extract(ctx)
+
+	l.Info("waiting for work....")
+
+	var done bool
+
+	for !done {
+		select {
+		case <-ctx.Done():
+			return c.handleContextCancel(ctx)
+		default:
+		}
+
+		nextTask, err := c.tasks.Next(ctx)
 		if err != nil {
-			ctxzap.Extract(ctx).Error("error shutting down", zap.Error(err))
+			l.Error("error getting next task", zap.Error(err))
+			continue
 		}
-	}(c, ctx)
 
-	sigChan := make(chan os.Signal, 1)
-	signal.Notify(sigChan, os.Interrupt)
-	go func() {
-		for range sigChan {
-			cancel()
+		// No work for us to do, go to sleep for a bit and try again
+		if nextTask == nil {
+			l.Debug("no available tasks, going to sleep for 5 seconds")
+			select {
+			case <-ctx.Done():
+				return c.handleContextCancel(ctx)
+			case <-time.After(time.Second * 5):
```

Comment:
Unsure about the amount of time here, I think it could actually probably be a lot higher. We might want it to be configurable by the task manager? `Next()` could return the next time to check again instead of it being hardcoded.

Reply:
yes, this should be a lot higher

Reply:
well its complicated. maybe not.  something to think about for sure.